File: javaexamples/src/main/java/com/inappstory/javaexamples/tagsplaceholders/TagsPlaceholdersSample.java
package com.inappstory.javaexamples.tagsplaceholders;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.inappstory.javaexamples.R;
import com.inappstory.sdk.AppearanceManager;
import com.inappstory.sdk.InAppStoryManager;
import com.inappstory.sdk.exceptions.DataException;
import com.inappstory.sdk.stories.ui.list.StoriesList;

import java.util.ArrayList;
import java.util.HashMap;

public class TagsPlaceholdersSample extends AppCompatActivity {



    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_placeholders);
        findViewById(R.id.refresh).setOnClickListener(v -> {
            setTags();
            addTags();
            removeTags();
            setPlaceholders();
        });
        showStories();
    }

    private void removeTags() {
        ArrayList<String> removedTags = new ArrayList<>();
        removedTags.add("tag3");
        InAppStoryManager.getInstance().removeTags(removedTags);
    }

    private void addTags() {
        ArrayList<String> additionalTags = new ArrayList<>();
        additionalTags.add("tag3");
        additionalTags.add("tag4");
        InAppStoryManager.getInstance().addTags(additionalTags);
    }

    private void setTags() {
        ArrayList<String> tags = new ArrayList<>();
        tags.add("tag1");
        tags.add("tag2");
        InAppStoryManager.getInstance().setTags(tags);
    }

    private void setPlaceholders() {
        HashMap<String, String> placeholders = new HashMap<>();
        placeholders.put("replaceme1", "value1");
        placeholders.put("replaceme2", "value2");
        InAppStoryManager.getInstance().setPlaceholders(placeholders);
        InAppStoryManager.getInstance().setPlaceholder("replaceme2", "value3");
    }

    private void showStories() {
        StoriesList storiesList = findViewById(R.id.stories_list);
        storiesList.setAppearanceManager(new AppearanceManager());
        try {
            storiesList.loadStories();
        } catch (DataException e) {
            e.printStackTrace();
        }
    }

    private void getTagsAndPlaceholders() {
        InAppStoryManager.getInstance().getTags();

        InAppStoryManager.getInstance().getTagsString();

        InAppStoryManager.getInstance().getPlaceholders();
    }

    @Override
    protected void onDestroy() {
        InAppStoryManager.getInstance().setTags(null);
        InAppStoryManager.getInstance().setPlaceholders(new HashMap<>());
        super.onDestroy();
    }
}

File: kotlinexamples/src/main/java/com/inappstory/kotlinexamples/tagsplaceholders/TagsPlaceholdersSample.kt
package com.inappstory.kotlinexamples.tagsplaceholders

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.inappstory.kotlinexamples.R
import com.inappstory.sdk.AppearanceManager
import com.inappstory.sdk.InAppStoryManager
import com.inappstory.sdk.stories.api.models.ImagePlaceholderValue
import com.inappstory.sdk.stories.ui.list.StoriesList
import java.util.ArrayList
import java.util.HashMap

class TagsPlaceholdersSample : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_placeholders)
        findViewById<View>(R.id.refresh).setOnClickListener { v: View? ->
            setTags()
            addTags()
            removeTags()
            setPlaceholders()
            setImagePlaceholders()
        }
        showStories()
    }

    private fun removeTags() {
        val removedTags = ArrayList<String>()
        removedTags.add("tag3")
        InAppStoryManager.getInstance().removeTags(removedTags)
    }

    private fun addTags() {
        val additionalTags = ArrayList<String>()
        additionalTags.add("tag3")
        additionalTags.add("tag4")
        InAppStoryManager.getInstance().addTags(additionalTags)
    }

    private fun setTags() {
        val tags = ArrayList<String>()
        tags.add("tag1")
        tags.add("tag2")
        InAppStoryManager.getInstance().tags = tags
    }


    private fun setPlaceholders() {
        val placeholders = HashMap<String, String>()
        placeholders["replaceme1"] = "value1"
        placeholders["replaceme2"] = "value2"

        InAppStoryManager.getInstance().setPlaceholders(placeholders)
        InAppStoryManager.getInstance().setPlaceholder("replaceme2", "value3")
    }

    private fun setImagePlaceholders() {
        val placeholders = hashMapOf<String, ImagePlaceholderValue>()
        placeholders["replaceme1"] = ImagePlaceholderValue.createByUrl("https://www.gstatic.com/webp/gallery/1.webp")
        InAppStoryManager.getInstance().setImagePlaceholders(placeholders)
        InAppStoryManager.getInstance()
            .setImagePlaceholder("replaceme2", ImagePlaceholderValue.createByUrl("https://www.gstatic.com/webp/gallery/2.webp"))
    }

    private fun showStories() {
        val storiesList = findViewById<StoriesList>(R.id.stories_list)
        storiesList.appearanceManager = AppearanceManager()
        storiesList.loadStories()
    }

    private val tagsAndPlaceholders: Unit
        private get() {
            InAppStoryManager.getInstance().tags
            InAppStoryManager.getInstance().tagsString
            InAppStoryManager.getInstance().placeholders
            InAppStoryManager.getInstance().imagePlaceholdersValues
        }

    override fun onDestroy() {
        InAppStoryManager.getInstance().tags = null
        InAppStoryManager.getInstance().placeholders = hashMapOf()
        super.onDestroy()
    }
}